File: src/ambiguity/constraints.py
"""Constraint taxonomy types (CDL — Constraint Detection Layer)."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum

from .parser import ParseResult


class ConstraintType(Enum):
    TECHNICAL = "technical"
    FINANCIAL = "financial"
    TEMPORAL = "temporal"
    HUMAN = "human"
    SEMANTIC = "semantic"
    INTEGRATION = "integration"
    EMERGENT = "emergent"


class ConstraintCategory(Enum):
    HARD = "hard"
    SOFT = "soft"
    FALSE = "false"


class ConstraintLevel(Enum):
    FEDERATION = "federation"
    GUILD = "guild"
    AGENT = "agent"
    ENTRY_POINT = "entry_point"


@dataclass
class Constraint:
    raw_text: str
    constraint_type: ConstraintType
    constraint_category: ConstraintCategory
    quantifiable: bool = False
    relaxable: bool = False
    priority: int = 5


def _categorize_by_kind(kind: str) -> ConstraintCategory:
    hard_kinds = {"requirement", "exact", "dependency", "negation"}
    false_kinds = {"assumption"}
    if kind in hard_kinds:
        return ConstraintCategory.HARD
    if kind in false_kinds:
        return ConstraintCategory.FALSE
    return ConstraintCategory.SOFT


def _type_for_kind(kind: str) -> ConstraintType:
    kind_type_map: dict[str, ConstraintType] = {
        "exact": ConstraintType.SEMANTIC,
        "requirement": ConstraintType.TECHNICAL,
        "negation": ConstraintType.SEMANTIC,
        "dependency": ConstraintType.INTEGRATION,
        "assumption": ConstraintType.EMERGENT,
        "conditional": ConstraintType.SEMANTIC,
    }
    return kind_type_map.get(kind, ConstraintType.TECHNICAL)


def upgrade_constraints(result: ParseResult) -> list[Constraint]:
    """Upgrade parser's raw constraint kind strings to typed Constraint objects."""
    typed: list[Constraint] = []
    for kind in result.constraints:
        ctype = _type_for_kind(kind)
        cat = _categorize_by_kind(kind)
        quant = kind in ("exact",)
        relax = kind in ("conditional",)
        typed.append(
            Constraint(
                raw_text=kind,
                constraint_type=ctype,
                constraint_category=cat,
                quantifiable=quant,
                relaxable=relax,
            )
        )
    return typed


@dataclass
class ConstraintAnalysis:
    typed: list[Constraint] = field(default_factory=list)
    hard_count: int = 0
    soft_count: int = 0
    false_count: int = 0
    missing_types: list[ConstraintType] = field(default_factory=list)

    @classmethod
    def from_parse(cls, result: ParseResult) -> ConstraintAnalysis:
        typed = upgrade_constraints(result)
        hard = sum(1 for c in typed if c.constraint_category == ConstraintCategory.HARD)
        soft = sum(1 for c in typed if c.constraint_category == ConstraintCategory.SOFT)
        false = sum(1 for c in typed if c.constraint_category == ConstraintCategory.FALSE)
        found_types = {c.constraint_type for c in typed}
        missing = [t for t in ConstraintType if t not in found_types]
        return cls(typed=typed, hard_count=hard, soft_count=soft, false_count=false, missing_types=missing)

File: ts/src/constraints.ts
/** Constraint taxonomy types (CDL — Constraint Detection Layer). */

import type { ParseResult } from "./parser.js"

export enum ConstraintType {
  TECHNICAL = "technical",
  FINANCIAL = "financial",
  TEMPORAL = "temporal",
  HUMAN = "human",
  SEMANTIC = "semantic",
  INTEGRATION = "integration",
  EMERGENT = "emergent",
}

export enum ConstraintCategory {
  HARD = "hard",
  SOFT = "soft",
  FALSE = "false",
}

export enum ConstraintLevel {
  FEDERATION = "federation",
  GUILD = "guild",
  AGENT = "agent",
  ENTRY_POINT = "entry_point",
}

export interface Constraint {
  rawText: string
  constraintType: ConstraintType
  constraintCategory: ConstraintCategory
  quantifiable: boolean
  relaxable: boolean
  priority: number
}

const HARD_KINDS = new Set(["requirement", "exact", "dependency", "negation"])
const FALSE_KINDS = new Set(["assumption"])

function categorizeByKind(kind: string): ConstraintCategory {
  if (HARD_KINDS.has(kind)) return ConstraintCategory.HARD
  if (FALSE_KINDS.has(kind)) return ConstraintCategory.FALSE
  return ConstraintCategory.SOFT
}

const KIND_TYPE_MAP: Record<string, ConstraintType> = {
  exact: ConstraintType.SEMANTIC,
  requirement: ConstraintType.TECHNICAL,
  negation: ConstraintType.SEMANTIC,
  dependency: ConstraintType.INTEGRATION,
  assumption: ConstraintType.EMERGENT,
  conditional: ConstraintType.SEMANTIC,
}

function typeForKind(kind: string): ConstraintType {
  return KIND_TYPE_MAP[kind] ?? ConstraintType.TECHNICAL
}

export function upgradeConstraints(result: ParseResult): Constraint[] {
  const typed: Constraint[] = []
  for (const kind of result.constraints) {
    const ct = typeForKind(kind)
    const cat = categorizeByKind(kind)
    const quant = kind === "exact"
    const relax = kind === "conditional"
    typed.push({
      rawText: kind,
      constraintType: ct,
      constraintCategory: cat,
      quantifiable: quant,
      relaxable: relax,
      priority: 5,
    })
  }
  return typed
}

export interface ConstraintAnalysis {
  typed: Constraint[]
  hardCount: number
  softCount: number
  falseCount: number
  missingTypes: ConstraintType[]
}

export function constraintAnalysisFromParse(result: ParseResult): ConstraintAnalysis {
  const typed = upgradeConstraints(result)
  const hard = typed.filter((c) => c.constraintCategory === ConstraintCategory.HARD).length
  const soft = typed.filter((c) => c.constraintCategory === ConstraintCategory.SOFT).length
  const falseC = typed.filter((c) => c.constraintCategory === ConstraintCategory.FALSE).length
  const foundTypes = new Set(typed.map((c) => c.constraintType))
  const missing = Object.values(ConstraintType).filter((t) => !foundTypes.has(t))
  return {
    typed,
    hardCount: hard,
    softCount: soft,
    falseCount: falseC,
    missingTypes: missing,
  }
}
